src/solver/multilabelsoftmarginloss/forward_reduced_multilabelsoftmarginloss.cpp

```diff
@@ -0,0 +1,230 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include "miopen/miopen.h"
+#include <miopen/datatype.hpp>
+#include <miopen/kernel_build_params.hpp>
+#include <miopen/multilabelsoftmarginloss/invoke_params.hpp>
+#include <miopen/multilabelsoftmarginloss/solvers.hpp>
+#include <miopen/multilabelsoftmarginloss.hpp>
+#include <miopen/target_properties.hpp>
+#include <miopen/tensor_view_utils.hpp>
+
+#define LOCAL_SIZE_MULTILABELSOFTMARGINLOSS 256
+#define LOCAL_SIZE_REDUCE 256
+
+namespace miopen {
+
+namespace solver {
+
+namespace multilabelsoftmarginloss {
+
+bool MultilabelSoftMarginLossForward::IsApplicable(
+    const ExecutionContext& /*context*/,
+    const miopen::multilabelsoftmarginloss::ForwardProblemDescription& problem) const
+{
+    if(problem.GetiDesc().GetLengths()[1] > 24)
+        return false;
+    return true;
+}
+
+ConvSolution MultilabelSoftMarginLossForward::GetSolution(
+    const ExecutionContext& /*context*/,
+    const miopen::multilabelsoftmarginloss::ForwardProblemDescription& problem) const
+{
+    auto result = ConvSolution{miopenStatusSuccess};
+
+    auto xgrid = problem.GetiDesc().GetLengths()[0];
+
+    {
+        /* Phase 1: Calc loss for each element. */
+        auto dtype        = problem.GetiDesc().GetType();
+        size_t xlocalsize = LOCAL_SIZE_MULTILABELSOFTMARGINLOSS;
+        size_t xgridsize  = AlignUp(xgrid, xlocalsize);
+        size_t ylocalsize = 1;
+        size_t ygridsize  = 1;
+        size_t zlocalsize = 1;
+        size_t zgridsize  = 1;
+
+        auto kernel        = KernelInfo{};
+        kernel.kernel_file = "MIOpenMultilabelSoftMarginLoss.cpp";
+        kernel.kernel_name = "MultilabelSoftMarginLossForward2d";
+
+        const auto build_params = KernelBuildParameters{
+            {"MIOPEN_USE_FP16", static_cast<int32_t>(dtype == miopenHalf)},
+            {"MIOPEN_USE_FP32", static_cast<int32_t>(dtype == miopenFloat)},
+            {"MIOPEN_USE_BFP16", static_cast<int32_t>(dtype == miopenBFloat16)},
+            {"INPUT_TYPE",
```

Comment:
Similar to above. If you aren't expecting input and output types to differ, you can simply remove this and use `FLOAT` in the kernel

Reply:
Fixed in [this commit](https://github.com/moreh-dev/MIOpen/pull/33/commits/e9f959f5c1bfb65a1fb4033449671487b45c9bf0)